Plugin-Manager

boombuler opened the pull request:
Hi everyone I've started to implement a plugin manager.

for now it supports the following commands:
* `plugin install [pluginnames]` download and installs the given plugin and their dependencies
* `plugin remove [pluginnames]` uninstall the given plugins
* `plugin update` updates all plugins and their dependencies.

There should be at least one global "channel" which is nothing more than a list of plugin repositories.
I propose one repository per developer and a central channel hosted by zyedidia.

Examples for this:

* `https://www.boombuler.de/channel.json` --> a channel (currently only one repo)
* `https://www.boombuler.de/repo.json` --> a repository with currently only one package and one version.

Each package can have multiple versions with different dependencies. 
Each plugin.lua should have a variable VERSION which is used to determine if there are any updates.
Like this:

```lua
VERSION = "1.0.1"

-- rest of the plugin
```

I will add more documentation later.


What is still missing:
* Searching for plugins.

boombuler commented:
Btw I'm happy if someone helps me with this. And I posted it in this early state so everyone can provide feedback or even code to this.

zyedidia commented:
Looks great!

I'm having an issue when it tries to resolve dependencies. The Version variable is not necessarily set to an actual version number and this crashes when it is either set to `nightly` or `Unknown`.

onodera-punpun commented:
I'm not well versed in go, but I'm pretty sure you are using json, maybe consider using json5 for consistencies sake (the setting.json file also uses json5).

onodera-punpun commented:
Another suggestion is to make `micro -version` output something like

```
Version: nightly
Commit hash: ac98f21
Compiled on September 25, 2016

Installed plugins: foo, bar
``

boombuler commented:
Hi,

the "no actual version" problem needs to be addressed. I've also noticed that you set the version number to "v1.0.3" instead of "1.0.3". It's an issue but I don't think that is the problem that will stop this from working.

I will have a look at the json5 stuff.

boombuler commented:
Hi again,

I've switched the code to json5.
Also I've modified the make file. If it matches the current workflow (which I don't know) it would behave like this:

* If the current commit has a version tag like "v1.0.3" this would be the version
* If the current commit has another tag like "nightly" it searches for the last known "version tag" and increments the patch number. 
* If there is no tag on the current commit "dev" is used as label.

Examples:
If you are on tag "nightly" which is 75 commits ahead of version "v1.0.3" the resulting version would be "1.0.4-nightly.75".

If you are on a non-tag revision which is 30 commits ahead of version "v1.0.3" it results in "1.0.4-dev-30"

If you are on the revision which is tagged with "v1.0.4" the resulting version is "1.0.4"

To make this work the tag has to be set before building the release...

As a fallback in version parsing within the plugin manager the behavior is:
* If the version is no "SemVer" try to parse `"0.0.0-"+Version` for example: 0.0.0-nightly
* If that fails too just treat the version as "0.0.0"

boombuler commented:
Update: new command:
`plugin search [something you search]`

zyedidia commented:
The plan here is to have one central "channel" which will contain a list of repositories hosted by developers, each with their plugins listed there, and I assume there will be some sort of selection to be admitted to the central channel? Why not just have a central repository of plugins and select per-plugin instead of per-developer?

In other words, what's the point of "channels"?

boombuler commented:
Simple: Do you want to merge a pull request everytime a developer releases a new version?

zyedidia commented:
Ok I think this only needs a couple more improvements before it's done.

There needs to be a way for users to list the plugins they currently have installed. Maybe that would be a command like `> plugin list` or maybe there could be a `plugins.json` file where users could also add plugins.

There should also be a way for users to add their own third-party repositories and/or channels.

Also I was playing around with it, trying to install the snippets plugin (http://zbyedidia.webfactional.com/snippets.zip) and when that file is unzipped, it unzips into a directory called `snippets`, so the directory structure becomes `configDir/plugins/snippets/snippets/...` which is a problem. I'm not sure what the best way to fix this is though.

Also I think if the version is `0.0.0-Unknown` or something like that (built with `go build`) the plugin manager should install any plugin regardless of the micro version dependency. Since it seems like the version is going to become much more important now, I'll change the building from source instructions in the readme to recommend `make install` instead of `go get`.

Once this is done I'll make a channel in a Github repo and add some repositories to that.

boombuler commented:
> There needs to be a way for users to list the plugins they currently have installed. Maybe that would be a command like > plugin list or maybe there could be a plugins.json file where users could also add plugins.

`plugin list` should work now but I don't get what you want to do with a plugins.json...

> There should also be a way for users to add their own third-party repositories and/or channels.

the channels and optionally additional repositories are now loaded from the settings. (Though it can't be set via command since the command doesn't accept json arrays...)

> Also I was playing around with it, trying to install the snippets plugin (http://zbyedidia.webfactional.com/snippets.zip) and when that file is unzipped, it unzips into a directory called snippets, so the directory structure becomes configDir/plugins/snippets/snippets/... which is a problem. I'm not sure what the best way to fix this is though.

Should be fixed. If all files in the zip are in the same directory, that directory is skipped. So it should even work if you zip a "release" directory or something like that.

> Also I think if the version is 0.0.0-Unknown or something like that (built with go build) the plugin manager should install any plugin regardless of the micro version dependency. Since it seems like the version is going to become much more important now, I'll change the building from source instructions in the readme to recommend make install instead of go get.

If the core version is not a valid version number, the requirements for the core are not evaluated anymore.

zyedidia commented:
Great.

I've been thinking that maybe it would be better if each repository only contained one package. That way developers would handle their own versioning and stuff, but being "official" is on a per package basis instead of per developer. It's confusing, but maybe it would be best to rename 'channels' to 'repositories' and 'repositories' to simply 'packages' or 'plugins'. Then a user would have list of 'repositories' in their settings (there would only be one 'official' repository) which would just be a list of links to `package.json` files.

Also, a little note, for json5 you should now use `github.com/zyedidia/json5` instead of `github.com/ yosuke-furukawa/json5`.

Let me know what you think. I think this is almost done 🎉 .

onodera-punpun commented:
>plugin list should work now but I don't get what you want to do with a plugins.json...

Wouldn't this make it easy to copy paste the micro dotfiles and have the exact same micro setup, unless the plugin manager saves the installed plugins somewhere.

boombuler commented:
> I've been thinking that maybe it would be better if each repository only contained one package. That way developers would handle their own versioning and stuff, but being "official" is on a per package basis instead of per developer. It's confusing, but maybe it would be best to rename 'channels' to 'repositories,' and 'repositories' to simply 'packages' or 'plugins'. Then a user would have list of 'repositories' in their settings (there would only be one 'official' repository) which would just be a list of links to package.json (currently repo.json) files.

If handled that way it raises the question what kind of metadata should be included in the repository. If the whole description etc would be at the "root" it would only require one http request. 
To be honest I had a look at sublimes pm and how they handle it. Seems like they had started with one central repo and then added channels because of the large number of pull requests. Although I don't see micros userbase exploding at the moment. I'm not sure what is the best way at this topic.

Let me know what you think...

> Also, a little note, for json5 you should now use github.com/zyedidia/json5 instead of github.com/ yosuke-furukawa/json5.

Fixed...


> Wouldn't this make it easy to copy paste the micro dotfiles and have the exact same micro setup everywhere.

I would also just copy / sync the configDir...

zyedidia commented:
Ok I think this is ready to be merged. I'm going to make a couple adjustments after I merge and I'm also going to create a repository in https://github.com/micro-editor to hold the main channel.

I think at first I'm only going to allow repositories to contain one package, but if this becomes a problem we can change it.

Would it be alright if I put the snippets plugin in a repository in the micro-editor organization or would you like to host it yourself? Let me know.

I also just wanted to say thank you for all the great work you did in getting this working 👍 .

boombuler commented:
> Would it be alright if I put the snippets plugin in a repository in the micro-editor organization or would you like to host it yourself? Let me know.

Its currently hosted here: https://github.com/boombuler/microsnippets but feel free to fork that repo. I will put my repo down once its forked.

> I just wanted to say thank you for all the great work you did in getting this working 👍 .
No problem. Glad I could help...
